dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/reservedvalue/ValueGeneratorService.java

```diff
@@ -45,70 +48,49 @@
     {
-        List<String> generatedValues = new ArrayList<>();
-
         switch ( segment.getMethod() )
         {
-
         case SEQUENTIAL:
-
-            BigInteger maxValue = BigInteger.TEN.pow( segment.getParameter().length() );
-            List<Integer> generatedNumbers = sequentialNumberCounterStore
-                .getNextValues( textPattern.getOwnerUid(), key, numberOfValues );
-
-            boolean outOfValues = generatedNumbers.stream()
-                .anyMatch( n -> maxValue.intValue() <= n );
-
-            if ( outOfValues )
-            {
-                throw new ReserveValueException( "Unable to reserve value, no new values available." );
-            }
-
-            generatedValues.addAll(
-                generatedNumbers
-                    .stream()
-                    .map( n -> String.format( "%0" + segment.getParameter().length() + "d", n ) )
-                    .collect( Collectors.toList() ) );
-
-            break;
-
+            return generateSequentialValues( segment, textPattern, key, numberOfValues );
         case RANDOM:
+            return generateRandomValues( segment, numberOfValues );
+        default:
+            return List.of();
+        }
+    }
 
-            List<Future<List<String>>> resultList = new ArrayList<>();
-
-            ExecutorService executorService = Executors.newFixedThreadPool( 10 );
-
-            randomGeneratorService.setSegmentParameter( segment.getParameter() );
-
-            for ( int i = 0; i < numberOfValues; i++ )
-            {
-                Future<List<String>> result = executorService
-                    .submit( randomGeneratorService );
-                resultList.add( result );
-            }
-
-            for ( Future<List<String>> result : resultList )
-            {
-                generatedValues.addAll( result.get() );
-            }
-
-            executorService.shutdown();
-
-            break;
+    private List<String> generateSequentialValues( TextPatternSegment segment, TextPattern textPattern, String key,
+        int numberOfValues )
+        throws ReserveValueException
+    {
+        BigInteger maxValue = BigInteger.TEN.pow( segment.getParameter().length() );
+        List<Integer> generatedNumbers = sequentialNumberCounterStore
+            .getNextValues( textPattern.getOwnerUid(), key, numberOfValues );
 
-        default:
-            break;
+        boolean outOfValues = generatedNumbers.stream()
+            .anyMatch( n -> maxValue.intValue() <= n );
 
+        if ( outOfValues )
+        {
+            throw new ReserveValueException( "Unable to reserve value, no new values available." );
         }
 
-        return generatedValues;
+        return generatedNumbers
+            .stream()
+            .map( n -> String.format( "%0" + segment.getParameter().length() + "d", n ) )
+            .collect( Collectors.toList() );
+    }
+
+    private List<String> generateRandomValues( TextPatternSegment segment, int numberOfValues )
```

Comment:
Maybe I am not following the code in the right way but lets say we want to generate 10 random values, if we have variable `numberOfValues` equals 10, then will we have this code generating 10 random values 10 times? So 100 values?

Reply:
Good question. I also find this super confusing. From what I gathered each call to random value generation creates a chunk. I just lifted the size of that chunk `RANDOM_GENERATION_CHUNK` out so it becomes an argument. Then we might do this a number of times so yes, the total result are `numberOfValues * RANDOM_GENERATION_CHUNK` values. Which is just a simplified view because generation might also output 2x the chunk, 2x chunk +1 or 3x chunk size depending on the actual pattern. This odd logic was thankfully covered by tests so I am sure I kept it as it is.

Reply:
:+1: